packages/idempotency/src/persistence/PersistenceLayer.ts

```diff
@@ -1,24 +1,161 @@
+   * 
+   * @param data - the data payload that will be hashed to create the hash portion of the idempotency key
+   */
+  public async getRecord(data: unknown): Promise<IdempotencyRecord> {
+    const idempotencyKey: string = this.getHashedIdempotencyKey(data);
+
+    return this._getRecord(idempotencyKey);
+  }
+
+  /**
+   * Saves a record indicating that the function's execution is currently in progress
+   * 
+   * @param data - the data payload that will be hashed to create the hash portion of the idempotency key
+   */
+  public async saveInProgress(data: unknown): Promise<void> { 
+    const idempotencyRecord: IdempotencyRecord = 
+    new IdempotencyRecord(this.getHashedIdempotencyKey(data),
+      IdempotencyRecordStatus.INPROGRESS,
+      this.getExpiryTimestamp(),
+      undefined,
+      undefined,
+      undefined
+    );
+
+    return this._putRecord(idempotencyRecord);
+  }
+
+  /**
+   * Saves a record of the function completing successfully. This will create a record with a COMPLETED status
+   * and will save the result of the completed function in the idempotency record.
+   * 
+   * @param data - the data payload that will be hashed to create the hash portion of the idempotency key
+   * @param result - the result of the successfully completed function
+   */
+  public async saveSuccess(data: unknown, result: unknown): Promise<void> { 
+    const idempotencyRecord: IdempotencyRecord = 
+    new IdempotencyRecord(this.getHashedIdempotencyKey(data),
+      IdempotencyRecordStatus.COMPLETED,
+      this.getExpiryTimestamp(),
+      undefined,
+      JSON.stringify(result),
+      undefined
+    );
+
+    this._updateRecord(idempotencyRecord);
   }
-  public async saveInProgress(): Promise<void> { }
-  public async saveSuccess(): Promise<void> { }
 
   protected abstract _deleteRecord(record: IdempotencyRecord): Promise<void>;
   protected abstract _getRecord(idempotencyKey: string): Promise<IdempotencyRecord>;
   protected abstract _putRecord(record: IdempotencyRecord): Promise<void>;
   protected abstract _updateRecord(record: IdempotencyRecord): Promise<void>;
+
+  /**
+   * Generates a hash of the data and returns the digest of that hash
+   * 
+   * @param data the data payload that will generate the hash
+   * @returns the digest of the generated hash
+   */
+  private generateHash(data: string): string{
+    const hash: Hash = createHash(this.hashFunction);
+    hash.update(data);
+    
+    return hash.digest(this.hashDigest);
+  }
+
+  /**
+   * Getter for `envVarsService`.
+   * Used internally during initialization.
+   */
+  private getEnvVarsService(): EnvironmentVariablesService {
+    return this.envVarsService;
+  }
+
+  /**
+   * Creates the expiry timestamp for the idempotency record
+   * 
+   * @returns the expiry time for the record expressed as number of seconds past the UNIX epoch
+   */
+  private getExpiryTimestamp(): number {
+    const currentTime: number = Date.now() / 1000;
```

Comment:
We may be out of sync here, the dynamo persistence layer does a comparison against Date.now() to determine whether the idempotency key is expired. So right now we would be making a comparison of milliseconds vs seconds as our conditional write.

`  const idempotencyKeyExpired = '#expiry < :now';
    const notInProgress = 'NOT #status = :inprogress';
    const conditionalExpression = `${idempotencyKeyDoesNotExist} OR ${idempotencyKeyExpired} OR ${notInProgress}`;
    try {
      await table.put({ TableName: this.tableName, Item: item, ExpressionAttributeNames: { '#id': this.key_attr, '#expiry': this.expiry_attr, '#status': this.status_attr }, ExpressionAttributeValues: { ':now': Date.now(), ':inprogress': IdempotencyRecordStatus.INPROGRESS }, ConditionExpression: conditionalExpression });
    } catch (e){`

Reply:
The reason I did that is that the configuration item is expiresAfterSeconds